File: currency-exchange-rates/src/main/java/com/exchange/rates/repository/ExchangeRateRepository.java
package com.exchange.rates.repository;

import com.exchange.rates.model.ExchangeRate;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface ExchangeRateRepository extends JpaRepository<ExchangeRate, Long> {
    @Query(value = "SELECT * from exchange_rate ex " +
            "where ex.from_currency_id = (select id from currency where symbol=:fromCurrencySymbol) " +
            "and ex.to_currency_id = (select id from currency where symbol=:toCurrencySymbol)",
            nativeQuery = true)
    Optional<ExchangeRate> getExchangeRate(String fromCurrencySymbol, String toCurrencySymbol);
}

File: currency-exchange-rates-js/src/repository/exchange-rate-repository.js
module.exports = (pool) => ({
  findExchangeRate: async ({ fromCurrencySymbol, toCurrencySymbol }) => {
    const { rows: [ firstEntry ] } = await pool.query({
      name: 'fetch-exchange-rate',
      text: `SELECT * FROM EXCHANGE_RATE WHERE 
        FROM_CURRENCY_ID = (SELECT ID FROM CURRENCY WHERE SYMBOL = $1)
        AND TO_CURRENCY_ID = (SELECT ID FROM CURRENCY WHERE SYMBOL = $2)`,
      values: [ fromCurrencySymbol, toCurrencySymbol ]
    })

    return firstEntry
  }
})
